src/main/scala/org/constellation/trust/TrustRank.scala

```diff
@@ -0,0 +1,204 @@
+package org.constellation.trust
+import scala.util.Random
+import cats.implicits._
+
+/**
+  * UNFINISHED -- Work in progress
+  * Modified version of EigenTrust
+  * First difference is scores are not normalized 0 -> 1 but rather -1 to 1
+  * This captures interference effects among distant networks.
+  * In addition, transitive scores for negative trusted Nth neighbors are discarded,
+  * as they are untrustworthy and shouldn't factor into the calculations
```

Comment:
Do you happen to have a reference to read up on the changes away from the generic algorithm? 
For the mapping of (0,1) to (-1,+1), does this more than just setting the edge from trustworthy to non-trustworthy from 1/2 to 0?

I'm thinking of possible issues when you want to introduce reduction of trust by division (which will not work if the values are negative, as then this actually raises their trust), and interpretation related to start values and normalization as used in the example you link. And if, for some reason, one may use to query a weighted opinion of the nodes and wanted to use their (normalized) trusts as individual weight (as in `(trust1*vote1 + trust2*vote2) / (trust1 + trust2)` for find a mean of two options on numberal values), then one would need to rescale those again.)

Reply:
https://nlp.stanford.edu/pubs/eigentrust.pdf
http://theory.stanford.edu/~za/HonestEigenTrust/HonestEigenTrust.pdf
https://www.researchgate.net/publication/261093756_EigenTrust_Attack_Resilient_Trust_Management

Was trying to get something closer to EigenTrust++ (which doesn't appear to have a Java implementation.) They don't use the nearest neighbors matrix multiplication approximation, instead they do a transitive sum over neighbor hops. 

Re: -1 to 1 -- this is actually more important for conflict detection than facilitator rankings. I think we'll probably need to use both actually, because 0 to 1 shows aggregate influence across the network, but -1 to 1 includes information about who your neighbors distrust which influences voting. If you normalize from 0 to 1 and you know there's a malicious peer, or your friends know, you want to incorporate that information. In regular EigenTrust it just gets 'ignored' or washed out because of the product operator going to 0, but it's more beneficial to use that when there's a fork to determine which side is appropriate. 

Re: division of the trust, the only reason to divide should be to normalize the outward degrees so someone who distrust a massive number of people spreads the influence less, there may be cases where we actually want to use that information though, so it's tricky (i.e. applying blanket distrust based on IP data for instance,) -- probably need to handle as a special case. -- but the normalized addition function I used works for both positive and negative (try out normalizedIncrease with + - sides, it should be consistent)

Re: start values, the distance metric is just used to simulate modularity (i.e. not random connections,) they should be starting out with -1 to 1 fully, not 1/3 -- unless I'm misinterpreting.

Yes -- re: weighted query, I skipped some steps, but EigenTrust++ handles this, it's part of the way to determine convergence. I just used an amplification factor + regular reduction function, but in reality it should be properly rescaled in the reducer, which needs to be done to guarantee convergence. They also use neighbor similarity when merging scores (i.e. Jaccard,) which would help as well.

Re: TLDR -- the problem is that transitive Nth order neighbor connections tend to 'wash out' when bordering around 0. All of the effects are lost for distrust when you take the product of N*N. It's more desirable to spread distrust as well through the network, to capture that effect.

Reply:
Thanks for the references @ryle-goehausen . Most I had encountered, exept I didn't see the Asian on is on ResearchGate.

As I write Bayesian updating algorithms for a living in C++, I'm a bit familiar with trust models (e.g. "subjective logic" extension to Bayeianism). 
After reading the Constellation references, I had previousl reviewed the Guru paper linked therein and made a review on youtube about it. With that one, I am sceptical of the severity of pubishments in practice for bigger systems.

https://youtu.be/8LP0hnKgtNI

>which doesn't appear to have a Java implementation.

>Re: -1 to 1 

In the end, since you can always map back and forth between any normalization (including what the neutral value 0 between -1 and +1 corresponds to), you will be able to detect any conflict in either representation. I can't gauge at this point what the cost-benefit is, I'm just playing devils advocate (always :) to make sure that one doesn't oversee e.g. when the positivity of the score might play a crucial role in any algorithmic part of the algorithm, or it's addons that have been developed. 
As you say, the scale between 0 and +1 represents a positive accumulated trust, and thus I'm also immediatenly sceptical of whether it makes sense that the  spetrum of dis-trust (0 down to -1) is equally as refined as positive ones. In reality there will alsoways be far more positvie than negative players (it's a more descrete matter) - what do the ranges between -1 and 0 mean, then? And now, whatever math you develope in the future to tune the algorithm, if both trust and distrust are on the same interval, then you'll always tune both together and this might not be what you want. E.g. you might have a parallel distrst scale that is being looked at in terms of whether it crosses a particular bound. Well this is just ideas.
Btw. I've actually proposed PageRank algorithms of that kind for other planned projects in the crypto space, where one rates not nodes but users (how actively they use the platform, and how credible they are, for crypot loans etc.)

>they should be starting out with -1 to 1 fully, not 1/3

The unit test in the link has an example which uses 1/3, that's where I come from. I don't know how your variant here fares without normalization and I don't know if once under 1, the algorithm can come back to ever again assign 1 to someone.
In case you're interested, I could (in a time within two or three weeks) do another youtube video revieing the papers you listed. Not sure if it's of interest to you, but maybe it is and once I're read into it, I'd have a more educated opinion.

I had searched around and found that NEM (top 20 market cap blockchian project, it even mentions Eigentrust++ early in their Wikipedia entry) uses Java and advertised to implement Eigentrust++. However, I couldn't find correpsonding code, so I don't know how far they are or if they even started. Maybe there's something out there, however, and probably it's worth contacting the devs (also because they are Java people and maybe they have something to peek at - also in other parts of such projects).